### PR TITLE
[5.5] Add a simple transformer

### PR DESCRIPTION
**Pull request by eddymaloo**
❤️ 

Original PR https://github.com/laravel/framework/pull/14357

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Seems reasonable. :)

File: src/Illuminate/Support/Fluent.php
Comment (GrahamCampbell) [inline src/Illuminate/Support/Fluent.php]: cs :P

File: src/Illuminate/Support/Fluent.php
Comment (GrahamCampbell) [inline src/Illuminate/Support/Fluent.php]: and here :)

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by eddymaloo** [comment]
PR updated. Thanks

**Comment by robclancy** [comment]
Will this account for relationships? Doesn't look like it? One of the reasons for a transformer is because of the dynamic relationships. If I use a policy that accesses a `user` relationship in one route and in another I don't use that policy I will now have the `user` in one response and not the other and this PR doesn't solve that. Nor does this PR handle nesting for said relationships.

I think `transform` shouldn't return the model but a new object that contains exactly what you want. And if you did that you would then be able to do stuff like...

```php
return $comment->transform(function ($comment) {
    return [
        'id' => $comment->id,
        'body' => $comment->body,
        'user' => $user->transform(function ($user) {
            'id' => $user->id,
            'name' => $user->name,
        }),
    ];
});
```

And then would show a clear point where you would want to refactor so the model itself defined this stuff so user could instead be something like `$user->someMethodUsingTransform()`. But then using the name `transform` is also wrong as you aren't so much transforming as generating a response.

We have used custom made transformers at work for a while and do things more fluent like (chain appends, casts, relationships, permissions etc) extending on what eloquent does. I think a better syntax would be using arrays which would then do some of the things eloquent already does.

```php
return $comment->transform([
  'visible' => ['id', 'body', 'draft'],
  'casts' => ['draft' => 'bool'],
  'appends' => ['user.emailHash'],
  'relationships' => ['user.profile'],
  'permissions/policies/whatever' => ['edit', 'delete'],
]);
```

Example response from that:
```
{
  id: 1,
  body: 'test comment',
  draft: false,
  canEdit: true,
  canDelete: false,
  user: {
    id: 4,
    name: 'Robbo',
    emailHash: '23423047we0esd',
  }
}
```

This would do nesting with the `.` like `with` does and account for extra things. The internal functionality would mainly just be setting things like `visible` to the models `protected $visible`. Relationships and permissions would have to be done a little different. For permissions you don't want to lose the loaded relationships but you want to specify which ones are returned with `toArray`. 

This is what we essentially do internally already with traits which also includes collections so we can nest in collections too. But the problem with ours is using fluent methods which is getting a bit cumbersome on our more complex requests and I plan to refactor to using the array method above.

**Comment by kamui545** [comment]
@robclancy According to the original PR relationship is supported.

```php
<?php 

namespace App\Transformers;

use Illuminate\Support\Transformer;
use App\Transformers\ProjectTransformer;

class UserTransformer extends Transformer
{
   public function handle(User $user) 
   {
      return [
         'email' => $user->email,
         'fullname' => $user->firstname . ' ' . $user->lastname,
         'projects' => $user->projects->transform(new ProjectTransformer());
      ];
   }
}

return App\User::all()->transform(new UserTransformer());
```

> If I use a policy that accesses a user relationship in one route and in another I don't use that policy I will now have the user in one response and not the other and this PR doesn't solve that.

I don't see any limitation for that, or maybe i just don't understand your issue.

This feature is simple and yet flexible, your implementation sounds like too complicated and specific to your use cases.

**Comment by robclancy** [comment]
I was looking at the diff and I see nothing like that.

**Comment by crynobone** [comment]
It's not a full solution, but rather enable it to be done. If even this small changes is so hard to be accepted, having a full solution seem to be the impossible, as everyone has own style of transforming data.

**Comment by GrahamCampbell** [comment]
Yeh, maybe it is best to omit this from the core?

File: src/Illuminate/Pagination/AbstractPaginator.php
Comment (GrahamCampbell) [inline src/Illuminate/Pagination/AbstractPaginator.php]: cs

File: src/Illuminate/Database/Eloquent/Model.php
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: cs

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by eddymaloo** [comment]
> Yeh, maybe it is best to omit this from the core?

No, it is not, I think you misunderstood.

CS fixed.

**Comment by crynobone** [comment]
> Yeh, maybe it is best to omit this from the core?

I'm no longer pursuing this, but don't mistaken it with amount of demands that were asking on the early PR.

Based on @robclancy suggestion.

```php
return $comment->transform([
  'visible' => ['id', 'body', 'draft'],
  'casts' => ['draft' => 'bool'],
  'appends' => ['user.emailHash'],
  'relationships' => ['user.profile'],
  'permissions/policies/whatever' => ['edit', 'delete'],
]);
```

This is beyond what `Illuminate\Support\Collection::transform()` features, his suggestion may not work well with Laravel unless "Laravel Contributors" looking for a huge breaking changes in say `5.5`.

**Comment by fernandobandeira** [comment]
This + maybe an abstract class should be enough, this implementation is pretty flexible, we can extend it in our apps however we want, we just need the base methods.

The abstract class used on @kamui545 example:

```php
<?php

namespace Illuminate\Support;

abstract class Transformer 
{
    public function __invoke(...$parameters)
    {
        return $this->handle(...$parameters);
    }
}
```

It's pretty simple and extendable, should cover almost all needs...
We can handle the visibility, appends, etc as suggested by @robclancy on our own app afterwards...

**Comment by robclancy** [comment]
This solution is incomplete as explained...

**Comment by robclancy** [comment]
And nothing I have said would be a breaking change.

File: src/Illuminate/Database/Eloquent/Model.php
Comment (CristianLlanos) [inline src/Illuminate/Database/Eloquent/Model.php]: Why should the model be wrapped with Fluent? 🤔

**Comment by eddymaloo** [comment]
@CristianLlanos See the original PR here https://github.com/laravel/framework/pull/14357

It is mostly to take advantage of the Jsonable/Arrayable implementation

**Comment by DCzajkowski** [comment]
I think using "full blown Fractal" makes way more sense since it has object-based transformers, makes responses etc. See https://github.com/spatie/laravel-fractal.

**Comment by robclancy** [comment]
Fractal is terrible lol. Slow, tedious and just not coded well. Any custom
solution will be better.

If fractal was good this wouldn't even be considered.

**Comment by tillkruss** [comment]
Unless relationships are supported, I agree with @GrahamCampbell to omit this from core. Maybe this could be solved as a framework-specific package (as an alternative to Fractal)?

**Comment by crynobone** [comment]
> Unless relationships are supported,

https://github.com/laravel/framework/pull/17565#issuecomment-275603653 isn't this enough for that. Or are you asking for auto relationship resolution, in that case would you see that it would be useful, since for version API you way want to hides, appends, mutate the attributes differently.

**Comment by deleugpn** [comment]
What's the take on this? Will we ever have built-in functionalities that replaces the need for full fractal? Seems like an old discussion with a bunch of PRs closed and no final decision.

**Comment by tillkruss** [comment]
@crynobone: Ah, could you give an example how this would work on a model with a relationship?

**Comment by crynobone** [comment]
In above example relates to `User::with('projects')->all();`

**Comment by adiachenko** [comment]
Can someone elaborate why this PR is stuck in processing for months? 

Seems like a really elegant solution to me that will enable people to build their custom response transformers upon existing framework facilities rather than bothering with Fractal.

**Comment by adiachenko** [comment]
> One of the reasons for a transformer is because of the dynamic relationships. If I use a policy that accesses a user relationship in one route and in another I don't use that policy I will now have the user in one response and not the other and this PR doesn't solve that. Nor does this PR handle nesting for said relationships.

@robclancy What you're saying is already supported by this PR. I feel like you're just jumping to conclusions based on your previous experience with transformers. 

Here is quick and dirty example of dynamic includes:

```
/**
 * Show specified volume
 *
 * @param Request $request
 * @param Post $post
 * @return \Illuminate\Http\JsonResponse
 */
public function show(Request $request, Post $post)
{
    if ($request->include == 'comments') {
        $post->load('comments');
    }

    $data = $post->transform(function ($post) {
        $result = [
            'title' => $post->title,
        ];

        if ($post->relationLoaded('comments')) {
            $result = array_merge($result, [
                'comments' => $post->comments->transform(function ($comment) {
                    return [
                        'body' => $comment->body
                    ];
                })
            ]);
        }

        return $result;
    });

    return response()->json($data);
}
```

Transformers shouldn't query relationships anyway.

**Comment by robclancy** [comment]
You clearly didn't understand a thing I said. And then go on about stuff that is obvious to anyone who uses Laravel.

**Comment by adiachenko** [comment]
@robclancy Well, maybe you're just pushing too much responsibility onto what should be the simple model transformation. This attitude is kind of the whole reason why Fractal became `slow, tedious and just not coded well` (in your own words). Don't swallow more than you can chew. This PR does one thing and does it well.

**Comment by robclancy** [comment]
Fractal was slow, tedious and not coded well from day 1 what are you even
talking about. Go fanboi somewhere else.

**Comment by adiachenko** [comment]
I don't even care anymore. I'll just use what I want and let you decide whatever business you had here discussing 20 lines of code for the past 9 months.

**Comment by robclancy** [comment]
lmao this guy

**Comment by deleugpn** [comment]
@robclancy lmao you, man. You're the one making a fool of yourself.

**Comment by robclancy** [comment]
Yes I'm, such a fool because I can't understand basic transformer usage and
why this PR isn't merged. Got me.

**Comment by crhayes** [comment]
I also don't understand why this needs to be in the model. 

What's the advantage of `$user->transform(new UserTransformer())` over `(new UserTransformer($user))->transform()`?

There are already a ton of libraries that do the latter. Why introduce even more functionality into the (already bloated) model?

Also @adiachenko, to play devil's advocate, the issue you're describing re: n+1 issues is not specific to Fractal. It's the fact that accessing a relationship as a property will automatically generate a query to fetch the related resource(s). You would have the same issues with your code if you didn't do the `relationLoaded` check (likewise, you can do this check in Fractal to prevent n+1 problems).

**Comment by crhayes** [comment]
Also, instead of:

```php
public function show(Request $request, Post $post)
{
    if ($request->include == 'comments') {
        $post->load('comments');
    }

    $data = $post->transform(function ($post) {
        $item = [
            'title' => $post->title,
        ];

        if ($post->relationLoaded('comments')) {
            $item['comments'] = $post->comments->transform(function ($comment) {
                return [
                    'body' => $comment->body
                ];
            });
        }

        return $item;
    });

    return response()->json($data);
}
```

Why not just:
```php
public function show(Request $request, Post $post)
{
    if ($request->include == 'comments') {
        $post->load('comments');
    }

    $data = [
        'title' => $post->title,
    ];

    if ($post->relationLoaded('comments')) {
        $data['comments'] = $post->comments->map(function ($comment) {
            return [
                'body' => $comment->body
            ];
        });
    }

    return response()->json($data);
}
```

**Comment by adiachenko** [comment]
@crhayes Yeah, I get that, but Fractal also does this stuff: `?include=comments:limit(5|1):order(created_at|desc)`. Omitting the issue why transformer even does query parsing (ParamBag is implicitly  [created for you](https://github.com/thephpleague/fractal/blob/master/src/Manager.php#L158) which is kind of a bloat), this directly encourages users of this package to use sub-optimal shortcuts instead of eager loading. Also, Fractal is pretty powerful but most of it's implementation is half-assed (e.g. `JsonApiSerializer` doesn't even conform the spec in what little it can do).

This is why I was on the fence about @robclancy pushing more and more ideas onto his would-be *complete transformation layer*. It's not really needed (by which I mean *it doesn't require integration with framework facilities, you can build that yourself*) and will only make the PR suck just as bad as Fractal. Simple and minimal is what we need right now. 

One of the parts I like the most about this is direct integration with Laravel's paginator. Otherwise, you're now required to do stuff like this (which is verbose and kinda ugly):

```
$paginator->setCollection(
    $paginator->getCollection()->transform($transformer)
);
```
Anyway, I don't want to turn this into polemic so please excuse me from further substantial comments. If project maintainers feel like this shouldn't be in the core, fine. I just wish some people weren't so negative about it.

**Comment by crhayes** [comment]
P.S. I agree with what you're saying about Fractal. We are using it at work and thinking about moving away from it.

I'm working on a little project right now and testing out https://github.com/tobscure/json-api. So far I like it. There's actually a PR open that will simplify the implementation even further (and apparently is 2-3x faster).

**Comment by dan-har** [comment]
You can checkout the package [presentit](https://github.com/dan-har/presentit) and the [laravel adapter](https://github.com/dan-har/presentit-laravel). 

It is a full solution to transform nested data structures with callable transformers and transformer classes as well as hiding keys without the need of if statements. In laravel it supports nested models, collections and relations.

It is a package I've published, and it was built out of the need to handle transformation of many entities for apis.

**Comment by deleugpn** [comment]
@dan-har do you know if it runs well with Dingo API?

**Comment by dan-har** [comment]
Not sure because I haven't used this package, but as long as you use any transformation package that is independent I think you will have any issues.

**Comment by crynobone** [comment]
@dan-har everyone can build this as a package, but why would you need 548 line of code when this can be done with just 54 (including comments and tests).

screen shot 2017-03-23 at 2 18 56 pm

**Comment by dan-har** [comment]
It comes down to what are the needs you defined and what interface you try to build. In your function you achieved nested closure transformers - if its good for what many project needs it is great. 

But there are many projects that might require fluent interface that handles code reuse for nested transformers and also reduces the number of if statements (there are many of those) inside the transformers. To achieve that it takes more code to write and I don't see any problem with that. 

Anyways I could not find out there a package that answers these requirements.

**Comment by crynobone** [comment]
With this PR being accepted:

```php
Present::item($user)->with(function (User $user) {
    return [
        'id' => $user->id,
        'name' => $user->first_name . " " . $user->last_name,
        'image' => $user->image ?: Present::hidden(),
        'friends' => Present::each($user->friends)->with(function (User $friend) {
            return [
                //...
            ]
        })
    ];
})->show();
```

Becomes

```php
$user->transform(function ($user) {
    return [
        'id' => $user->id,
        'name' => $user->first_name . " " . $user->last_name,
        'image' => $user->image ?: null,
        'friends' => $user->friends->transform(function (User $friend) {
            return [
                //...
            ];
        })
    ];
}); // this return Arrayable and Jsonable where you can just output it directly to a Response.
```

And that with 54 LoC added to the core, nothing else.

**Comment by dan-har** [comment]
This is correct, but just note that the are cases where you still want to show null values while other keys are hidden and then you have to write a callback function and pass it to the array_filter or use if statements.

**Comment by crynobone** [comment]
> but just note that there are cases where you still want to show null values while other keys are hidden and then you have to write a callback function and pass it to the array_filter or use if statements.

This can be done on a base `Transformer` class that everyone can has the freedom to define how they want to do it. 

For me personally it doesn't make sense to have partial hidden value since the client (frontend/API consumer) then would need to have tonne of `if`/`else` to cater all this scenario.

**Comment by adiachenko** [comment]
> I guess that code example makes sense, considering collections are immutable. I don't think you should be able to call a method that mutates the underlying collection in the paginator object.

btw, [here is the discussed code](https://github.com/laravel/framework/pull/17565#issuecomment-287623574).

@crhayes Collections are not immutable. So why should paginator? 

There is already [transform](https://laravel.com/docs/5.4/collections#method-transform) method for collections in the core and that's why this PR works much better than earlier proposed implementations (e.g. having transformers as classes instead of a callable) that pass data all over the place.

With this, you'll have the ability to perform transformations inline with:
- Collections (already available in the core)
- Models
- Paginator instance (can't be extracted into separate package because `paginate` method on query builder is hardcoded to `LengthAwarePaginator` and the latter isn't macroable)

Funnily, the core will actually look more consistent with this PR merged.

**Comment by crhayes** [comment]
@adiachenko Collections are immutable. It's in the documentation.

image

I don't believe every method is, but most are. If you look at the code base most of the Collection operations `return new static(...)`, which means they are immutable as they are returning a new instance.

**Comment by crynobone** [comment]
`Collection::transform()` isn't immutable, you can see the code <https://github.com/laravel/framework/blob/5.4/src/Illuminate/Support/Collection.php#L1254-L1265>

**Comment by adiachenko** [comment]
@crhayes You can't be *mostly* immutable. That defies the whole point of immutability. You either immutable or you not. 😄 Why makes you think that having transform method for Eloquent Collections is fine, but Eloquent Models or Paginator (which is pretty much just a wrapped collection) are somehow special?

**Comment by crhayes** [comment]
@adiachenko Unfortunately for your argument that is exactly the case. Most of the methods *are* immutable. There are a couple exceptions (as I mentioned). If you take issue with that then I suggest you open a PR to update the official Laravel documentation 😄 .

I think the main advantage of this proposal is probably being able to take advantage of `Jsonable` (which I know was mentioned previously in this thread). That sounds reasonable to me.

I'm not really that passionate about this issue (whether or not it should go in). I just wanted to bring my opinion to the discussion, so thanks for the discussion :)

**Comment by adiachenko** [comment]
@crhayes I don't have any issues with documentation. It's just that immutability doesn't really apply to this discussion because the thing isn't immutable to begin with. Whether you mutate underlying data by calling `setCollection` or `transform` is pretty much a difference in aesthetics only. 

We're just deviate from the real question at hand here which is **exactly why this shouldn't be merged**? It's like literally 10 lines of code that brings consistency to already existing API and isn't affecting old projects. Going by Laravel history, it can even go live with the upcoming patch release. It just pains me to see the lack of clear and unambiguous response on the matter.

**Comment by dan-har** [comment]
@crynobone The ```Hidden``` object suppose to indicates a key that should not be presented to the client, 
and there is no need to implement anything in the client -  You gave a good example of how to implement it with callbacks.

 The purpose for this object is to clean the transformers from 'visibility' if statement, for example lets say that only post owner is allowed to see the number of views of his post, so a possible implementation would be:

```php
$user->transform(function (Post $post) use ($user) {
    $presentation = [
        'title' => $post->title,
        'text' => $post->text,
    ];

   if($post->isOwner($user)) {
         presentation['views'] = $post->viewCount();
   }

    return $presentation;
});
```
and with a Hidden object you can write (using your base transformer from above):

```php

class PostTransformer extends BaseTransformer 
{
     public function __construct(User $user) {
           $this->user = $user;
     }

     public function handle(Post $post) {
         return  [
             'title' => $post->title,
             'text' => $post->text,
             'views' => $post->isOwner($this->user) ? $post->viewCount() : new Hidden()
         ];
     };
}

$post->transform(new PostTransformer($user))
```

**Comment by taylorotwell** [comment]
I'm not ready to commit to merging this into the core at this time (or any other fractal API type package). However, I will definitely be revisiting all of this before the 5.5 release and will add this PR to my list of PRs to look over again when I get to that point. Thanks.

**Pull request closed without merging**
